flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/snapshot/RocksIncrementalSnapshotStrategy.java

```diff
@@ -403,85 +409,6 @@ private void cleanupIncompleteSnapshot(@Nonnull List<StateObject> statesToDiscar
 			}
 		}
 
-		private void uploadSstFiles(
-			@Nonnull Map<StateHandleID, StreamStateHandle> sstFiles,
-			@Nonnull Map<StateHandleID, StreamStateHandle> miscFiles) throws Exception {
-
-			// write state data
-			Preconditions.checkState(localBackupDirectory.exists());
-
-			FileStatus[] fileStatuses = localBackupDirectory.listStatus();
-			if (fileStatuses != null) {
-				for (FileStatus fileStatus : fileStatuses) {
-					final Path filePath = fileStatus.getPath();
-					final String fileName = filePath.getName();
-					final StateHandleID stateHandleID = new StateHandleID(fileName);
-
-					if (fileName.endsWith(SST_FILE_SUFFIX)) {
-						final boolean existsAlready =
-							baseSstFiles != null && baseSstFiles.contains(stateHandleID);
-
-						if (existsAlready) {
-							// we introduce a placeholder state handle, that is replaced with the
-							// original from the shared state registry (created from a previous checkpoint)
-							sstFiles.put(
-								stateHandleID,
-								new PlaceholderStreamStateHandle());
-						} else {
-							sstFiles.put(stateHandleID, uploadLocalFileToCheckpointFs(filePath));
```

Comment:
I think this logic of sst/placeholders/misc file separation belongs here.
`RocksDbStateDataTransfer` should be responsible only for the upload.

What if instead of calling `uploadLocalFileToCheckpointFs`, we fill `sstFilesToUpload` and `miscFilesToUpload` of `Map<StateHandleID, Path>` in this loop and then call `RocksDbStateDataTransfer.uploadStateFiles` for each map? This `uploadStateFiles` could return `Map<StateHandleID, StreamStateHandle>` which could be put into the final `sstFiles`/`miscFiles`.

I think `ConcurrentHashMap` is not needed then.